In [1]:
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import pandas as pd
from sklearn.utils import shuffle
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split

# Load data
csv_features = "./acsincome_ca_features.csv"
csv_labels = "./acsincome_ca_labels.csv"
features = pd.read_csv(csv_features)
labels = pd.read_csv(csv_labels)

# Separating features and labels, shuffling
X_all = features
Y_all = labels
X_all,Y_all = shuffle(X_all,Y_all, random_state=0)

# Only use first N samples to limit training time
sample = 0.2
num_samples = int(len(X_all)*sample)
X,Y = X_all[:num_samples], Y_all[:num_samples]

# Scaler for data standardization
scaler = StandardScaler()

# Separate train and test sets
X_train, X_test, Y_train, Y_test = train_test_split(X,Y, random_state=0, train_size=0.8)

X_train = scaler.fit_transform(X_train) 

# Random Forest

### Modèle

In [2]:
from sklearn.ensemble import RandomForestClassifier

RF_model_def = RandomForestClassifier()

RF_model_def.fit(X_train, Y_train)
print("Modèle avec paramètres par défaut", RF_model_def.score(X_test, Y_test))

/home/mecaliff/.conda/envs/appsupervise-tp/lib/python3.11/site-packages/sklearn/base.py:1474: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return fit_method(estimator, *args, **kwargs)


Modèle avec paramètres par défaut 0.4397598058004344


/home/mecaliff/.conda/envs/appsupervise-tp/lib/python3.11/site-packages/sklearn/base.py:486: UserWarning: X has feature names, but RandomForestClassifier was fitted without feature names
  warnings.warn(


Score modèle défaut : 10% : 0.5524
 
Score modèle défaut : 20% : 0.5868

Score modèle défaut : 50% : 0.5874

Score modèle défaut : 75% : 0.5845

Score modèle défaut : 100% : 0.5896

Paramètres défaut : param_def ={
    n_estimators : 100,
    max_depth : None,
    min_samples_split : 2,
    min_samples_leaf : 1
} 

## 1. Validation croisée

In [3]:
import warnings
warnings.filterwarnings("ignore")
#
# param_def ={
#     n_estimators : 100,
#     max_depth : None,
#     min_samples_split : 2,
#     min_samples_leaf : 1
# } 
RF_model = RandomForestClassifier(n_estimators=500, max_depth=100,
                        min_samples_split=40, min_samples_leaf=15)

RF_model_def.fit(X_train, Y_train)
print("Modèle avec paramètres 1", RF_model_def.score(X_test, Y_test))

Modèle avec paramètres 1 0.5900089434010477


## 2. Evaluation de la qualité du modèle en utilisant différentes métriques : accuracy, classificiation_report, confusion_matrix

In [32]:
for t in range(1, 5) :
    print(t)

1
2
3
4


## 3. Recherche des meilleurs hyperparamètres => Gridsearchcv

Sur un pourcentage croissant des données

In [4]:
from sklearn.model_selection import GridSearchCV

param_grid ={
    'n_estimators' :[100, 200, 300, 500],
    'max_depth' : [None, 30, 40, 50, 100],
    'min_samples_split' :[10, 20, 30, 40],
    'min_samples_leaf' :[4, 6, 8, 15]   
} 
grid= GridSearchCV(estimator= RandomForestClassifier(random_state=0),
                   param_grid=param_grid, scoring='accuracy', cv=5)

grid.fit(X_train, Y_train)

print(grid.best_score_)
print(grid.best_params_)

0.8143169191179129
{'max_depth': None, 'min_samples_leaf': 4, 'min_samples_split': 40, 'n_estimators': 100}


0.8143169191179129

{'max_depth': None, 'min_samples_leaf': 4, 'min_samples_split': 40, 'n_estimators': 100}